File: driveway_check_web/src/lib/identify/skeleton-extractor.ts
/** Approximate medial axis extraction for a 2D footprint.
 *
 * Ported from Part1_Identify/Core/SkeletonExtractor.cs.
 *
 *   1. Densely sample the boundary (exterior + holes) into points.
 *   2. Compute the Voronoi diagram of those points.
 *   3. Keep only edges whose BOTH endpoints fall strictly inside the polygon.
 *
 * There is no ridge-vertex output, so edges are recovered from the per-site cell
 * polygons: every internal Voronoi edge is shared by exactly two adjacent cells,
 * so collecting and deduping ring segments reconstructs the same ridge graph.
 *
 * Raw output is noisy — every reflex vertex produces a spurious short branch
 * toward it. See skeleton-pruning. */

import { SkeletonGraph, type Edge, type Pt } from "./skeleton-graph";
import {
	boundary_distance,
	coord,
	interior_locator,
	interior_rings,
	is_topology_failure,
	polygon_from_ring,
	voronoi_cells,
	type JtsCoordinate,
	type JtsGeometry,
	type JtsPolygon,
} from "./jts";
import { WidthProfile } from "./width-profile";

/** The Voronoi builder throws on certain sample configurations — a degeneracy in
 * the underlying triangulation, not a flaw in the shape. Tolerance does not fix
 * it (the full 0.001-1.0 range was tried). Nor does scaling the spacing alone: a
 * real T-shape failed every attempt up to +5% because uniform scaling preserves
 * the symmetric coincidences causing it. Per-point jitter perturbs points
 * independently and does break them, so both are applied. */
const MAX_ATTEMPTS = 12;
const SPACING_GROWTH = 1.0123;
const JITTER_FRACTION = 1e-4;
/** Two cell-ring vertices closer than this fuse into one skeleton node. */
const QUANTIZE = 1e-6;
/** Same floor as the Python prototype's _sample_boundary. */
const MIN_SAMPLES_PER_RING = 8;

/** Deterministic PRNG. .NET's Random(12345) stream cannot be reproduced in JS, so
 * the sequence differs from the C# — but it is stable run to run, which is what
 * the retry actually needs. Constructed once outside the loop so state carries
 * across attempts, matching the original. */
function mulberry32(seed: number)
{
	let a = seed >>> 0;
	return () =>
	{
		a = (a + 0x6d2b79f5) >>> 0;
		let t = Math.imul(a ^ (a >>> 15), 1 | a);
		t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
		return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
	};
}

function ring_length(coords: JtsCoordinate[]): number[]
{
	const cumulative = new Array<number>(coords.length).fill(0);
	for (let i = 1; i < coords.length; i++)
	{
		cumulative[i] = cumulative[i - 1] + Math.hypot(
			coords[i].x - coords[i - 1].x,
			coords[i].y - coords[i - 1].y
		);
	}
	return cumulative;
}

function point_at_length(coords: JtsCoordinate[], cumulative: number[], target: number): Pt
{
	// C# Array.BinarySearch returns the index on an exact hit and ~insertionIndex
	// otherwise; this reproduces the insertion index directly.
	let lo = 0;
	let hi = cumulative.length - 1;
	while (lo < hi)
	{
		const mid = (lo + hi) >> 1;
		if (cumulative[mid] < target) lo = mid + 1;
		else hi = mid;
	}
	const i = Math.min(Math.max(lo, 1), coords.length - 1);

	const segment = cumulative[i] - cumulative[i - 1];
	const t = segment > 1e-12 ? (target - cumulative[i - 1]) / segment : 0;
	return {
		x: coords[i - 1].x + t * (coords[i].x - coords[i - 1].x),
		y: coords[i - 1].y + t * (coords[i].y - coords[i - 1].y),
	};
}

function sample_ring(ring: JtsGeometry, spacing: number, out: Pt[]): void
{
	const coords = ring.getCoordinates();
	const cumulative = ring_length(coords);
	const length = cumulative[cumulative.length - 1];
	const n = Math.max(Math.trunc(length / spacing), MIN_SAMPLES_PER_RING);
	for (let i = 0; i < n; i++) out.push(point_at_length(coords, cumulative, (length * i) / n));
}

function sample_boundary(polygon: JtsPolygon, spacing: number): Pt[]
{
	const points: Pt[] = [];
	sample_ring(polygon.getExteriorRing(), spacing, points);
	for (const hole of interior_rings(polygon)) sample_ring(hole, spacing, points);
	return points;
}

export function extract_skeleton(polygon: JtsPolygon, sample_spacing: number): SkeletonGraph
{
	const jitter_random = mulberry32(12345);
	let cells: JtsPolygon[] = [];

	for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++)
	{
		let points = sample_boundary(polygon, sample_spacing * SPACING_GROWTH ** attempt);
		if (attempt > 0)
		{
			const jitter = sample_spacing * JITTER_FRACTION;
			points = points.map((p) => ({
				x: p.x + (jitter_random() - 0.5) * jitter,
				y: p.y + (jitter_random() - 0.5) * jitter,
			}));
		}
		// A single site yields no cells at all.
		if (points.length < 2) continue;
		try
		{
			// Real Coordinate instances, not plain objects: jsts calls .copy() on them
			// internally and a duck-typed {x,y} fails deep inside the triangulation.
			cells = voronoi_cells(points.map((p) => coord(p.x, p.y)));
			break;
		}
		catch (error)
		{
			// The last attempt rethrows deliberately. Swallowing it would hand
			// downstream an empty skeleton, which reads as "this footprint has no
			// medial axis" rather than as a failure.
			if (attempt === MAX_ATTEMPTS - 1 || !is_topology_failure(error)) throw error;
		}
	}

	const isInside = interior_locator(polygon);
	const node_index = new Map<string, number>();
	const nodes: Pt[] = [];
	const edge_set = new Set<string>();
	const edges: Edge[] = [];

	// Containment is the hot path — one lookup per cell-ring vertex, and vertices
	// are shared between adjacent cells. Memoising on the same key used for node
	// dedup removes tens of thousands of repeat tests.
	const containment = new Map<string, boolean>();
	const key = (c: JtsCoordinate) =>
		`${Math.round(c.x / QUANTIZE)},${Math.round(c.y / QUANTIZE)}`;

	const inside = (c: JtsCoordinate, k: string) =>
	{
		const cached = containment.get(k);
		if (cached !== undefined) return cached;
		const result = isInside(c);
		containment.set(k, result);
		return result;
	};

	const node_for = (c: JtsCoordinate, k: string) =>
	{
		const existing = node_index.get(k);
		if (existing !== undefined) return existing;
		const index = nodes.length;
		nodes.push({ x: c.x, y: c.y });
		node_index.set(k, index);
		return index;
	};

	for (const cell of cells)
	{
		const ring = cell.getExteriorRing().getCoordinates();
		for (let i = 0; i + 1 < ring.length; i++)
		{
			const c1 = ring[i];
			const c2 = ring[i + 1];
			const k1 = key(c1);
			const k2 = key(c2);
			// Discard anything touching the clip envelope or outside the footprint.
			if (!inside(c1, k1) || !inside(c2, k2)) continue;

			const i1 = node_for(c1, k1);
			const i2 = node_for(c2, k2);
			if (i1 === i2) continue;
			const edgeKey = i1 < i2 ? `${i1},${i2}` : `${i2},${i1}`;
			if (edge_set.has(edgeKey)) continue;
			edge_set.add(edgeKey);
			edges.push(i1 < i2 ? [i1, i2] : [i2, i1]);
		}
	}

	return new SkeletonGraph(nodes, edges);
}

/** Local width at every skeleton node: twice the distance to the boundary. */
export function compute_widths(polygon: JtsPolygon, graph: SkeletonGraph): WidthProfile
{
	const distanceTo = boundary_distance(polygon.getExteriorRing());
	const widths = new Map<number, number>();
	for (let i = 0; i < graph.nodes.length; i++)
	{
		const node = graph.nodes[i];
		widths.set(i, 2 * distanceTo(coord(node.x, node.y)));
	}
	return new WidthProfile(widths);
}

export { polygon_from_ring };

File: driveway_check_rhino/src/Part1_Identify/Core/SkeletonExtractor.cs
using System;
using System.Collections.Generic;
using NetTopologySuite.Geometries;
using NetTopologySuite.Geometries.Prepared;
using NetTopologySuite.Triangulate;

namespace DrivewayChecker.Identify.Core
{
    // Approximate medial axis (skeleton) extraction for a 2D polygon footprint.
    //
    // No RhinoCommon dependency by design -- operates on plain NTS Polygon/Coordinate
    // types, so it can be developed and tested standalone via `run_csharp` against
    // synthetic shapes, with no Rhino round-trip needed. Ported from
    // reference/core/skeleton.py, same method:
    //   1. Densely sample the polygon boundary (exterior + holes) into points.
    //   2. Compute the Voronoi diagram of those boundary points.
    //   3. Keep only edges whose both endpoints fall strictly inside the polygon.
    //      What survives approximates the medial axis.
    //
    // NTS has no direct scipy-Voronoi-style ridge-vertex output, so edges are
    // recovered from VoronoiDiagramBuilder's per-site cell polygons instead: every
    // internal Voronoi edge is shared by exactly two adjacent cells' boundary rings,
    // so collecting+deduping ring segments across all cells reconstructs the same
    // ridge graph scipy would hand back directly (verified live against NTS 2.5.0).
    //
    // Known limitation (not yet solved by extraction alone): raw output is noisy --
    // every reflex (concave) vertex of the input polygon produces a spurious short
    // branch toward it. See SkeletonPruning.
    public static class SkeletonExtractor
    {
        public static SkeletonGraph ExtractSkeleton(Polygon polygon, double sampleSpacing)
        {
            var gf = polygon.Factory;

            // NTS's classic overlay algorithm can throw TopologyException:
            // found non-noded intersection for certain sample-point
            // configurations -- a numerical degeneracy in the underlying
            // Delaunay triangulation, not a flaw in the input shape (found on
            // a real "staple"-shaped footprint with two close reflex
            // corners). VoronoiDiagramBuilder.Tolerance does NOT fix this --
            // tried the full range from 0.001 to 1.0, all failed identically.
            // A pure spacing-scale change isn't reliable on its own either:
            // found a real building footprint (a T-shape with a two-legged
            // notch) that still failed on every attempt up to a 5% spacing
            // change, because that scale change alone can preserve an exact
            // coincidence between two independently-sampled points on a
            // shape with round, symmetric proportions -- scaling everything
            // uniformly doesn't break that symmetry. Adding a small random
            // jitter to each sampled point (once spacing alone hasn't
            // worked) does break it, since it perturbs points independently
            // rather than as a group.
            const int maxAttempts = 12;
            var jitterRandom = new Random(12345);
            List<Coordinate> boundaryPts = null;
            GeometryCollection diagram = null;
            for (int attempt = 0; attempt < maxAttempts; attempt++)
            {
                boundaryPts = SampleBoundary(polygon, sampleSpacing * Math.Pow(1.0123, attempt));
                if (attempt > 0)
                {
                    double jitter = sampleSpacing * 1e-4;
                    for (int k = 0; k < boundaryPts.Count; k++)
                        boundaryPts[k] = new Coordinate(
                            boundaryPts[k].X + (jitterRandom.NextDouble() - 0.5) * jitter,
                            boundaryPts[k].Y + (jitterRandom.NextDouble() - 0.5) * jitter);
                }
                try
                {
                    var vb = new VoronoiDiagramBuilder();
                    vb.SetSites(boundaryPts);
                    diagram = vb.GetDiagram(gf);
                    break;
                }
                catch (TopologyException) when (attempt < maxAttempts - 1)
                {
                }
            }

            var prepared = PreparedGeometryFactory.Prepare(polygon);

            var nodeIndex = new Dictionary<(long, long), int>();
            var nodes = new List<Coordinate>();
            var edgeSet = new HashSet<(int, int)>();
            var edges = new List<(int, int)>();

            int GetOrAddNode(Coordinate c)
            {
                var key = Quantize(c);
                if (nodeIndex.TryGetValue(key, out var idx)) return idx;
                idx = nodes.Count;
                nodes.Add(c);
                nodeIndex[key] = idx;
                return idx;
            }

            for (int gi = 0; gi < diagram.NumGeometries; gi++)
            {
                if (diagram.GetGeometryN(gi) is not Polygon cell) continue;
                var ring = cell.ExteriorRing.Coordinates;
                for (int i = 0; i + 1 < ring.Length; i++)
                {
                    var c1 = ring[i];
                    var c2 = ring[i + 1];
                    if (!prepared.Contains(gf.CreatePoint(c1)) || !prepared.Contains(gf.CreatePoint(c2)))
                        continue; // discard: touches the clip envelope or falls outside the footprint

                    int i1 = GetOrAddNode(c1);
                    int i2 = GetOrAddNode(c2);
                    if (i1 == i2) continue;
                    var key = i1 < i2 ? (i1, i2) : (i2, i1);
                    if (edgeSet.Add(key)) edges.Add(key);
                }
            }

            return new SkeletonGraph(nodes, edges);
        }

        static (long, long) Quantize(Coordinate c, double precision = 1e-6)
            => ((long)Math.Round(c.X / precision), (long)Math.Round(c.Y / precision));

        static List<Coordinate> SampleBoundary(Polygon polygon, double spacing)
        {
            var pts = new List<Coordinate>();
            SampleRing(polygon.ExteriorRing, spacing, pts);
            foreach (var hole in polygon.InteriorRings)
                SampleRing(hole, spacing, pts);
            return pts;
        }

        // Evenly spaced samples along one ring, same `max(n, 8)` floor as the
        // Python prototype's _sample_boundary.
        static void SampleRing(LineString ring, double spacing, List<Coordinate> outPts)
        {
            double length = ring.Length;
            int n = Math.Max((int)(length / spacing), 8);
            var coords = ring.Coordinates;

            var cumulative = new double[coords.Length];
            for (int i = 1; i < coords.Length; i++)
                cumulative[i] = cumulative[i - 1] + Dist(coords[i - 1], coords[i]);

            for (int i = 0; i < n; i++)
                outPts.Add(PointAtLength(coords, cumulative, length * i / n));
        }

        static Coordinate PointAtLength(Coordinate[] coords, double[] cumulative, double target)
        {
            int i = Array.BinarySearch(cumulative, target);
            if (i < 0) i = ~i;
            i = Math.Clamp(i, 1, coords.Length - 1);

            double segLen = cumulative[i] - cumulative[i - 1];
            double t = segLen > 1e-12 ? (target - cumulative[i - 1]) / segLen : 0.0;
            double x = coords[i - 1].X + t * (coords[i].X - coords[i - 1].X);
            double y = coords[i - 1].Y + t * (coords[i].Y - coords[i - 1].Y);
            return new Coordinate(x, y);
        }

        static double Dist(Coordinate a, Coordinate b)
        {
            double dx = a.X - b.X, dy = a.Y - b.Y;
            return Math.Sqrt(dx * dx + dy * dy);
        }
    }
}
